UefiTestingPkg/FunctionalSystemTests/SmmPagingProtectionsTest/Smm/SmmPagingProtectionsTestSmm.c

```diff
@@ -231,13 +257,245 @@ SmmMemoryProtectionsTestInvalidRange (
+  if (Status == EFI_UNSUPPORTED) {
+    return EFI_UNSUPPORTED;
+  }
+  DEBUG ((DEBUG_ERROR, "[%a] - System proceeded through what should have been a critical failure!\n", __FUNCTION__));
+
+  return EFI_SECURITY_VIOLATION;
+}
+
+/**
+  Unauthorized MSR write test.
+  This test will execute platform code to attempt an MSR write operation that should be prevented in SMM
+
+  @retval     EFI_SECURITY_VIOLATION  Since this test is supposed to produce
+                                      a system crash, any sort of return value
+                                      should be considered a security violation.
+
+**/
+EFI_STATUS
+EFIAPI
+SmmMemoryProtectionsTestUnathorizedMsrWrite (
+  VOID
+  )
+{
+  EFI_STATUS Status;
+
+  DEBUG ((DEBUG_VERBOSE, "%a()\n", __FUNCTION__));
+
+  // Make sure that the required telemetry/handling is
+  // performed to get accurate test results.
+  EnableExceptionTestMode ();
+
+  DEBUG ((DEBUG_INFO, "[%a] - Attempting unauthorized MSR write.\n", __FUNCTION__));
+  Status = TestUnauthorizedMsrWrite();
+  if (Status == EFI_UNSUPPORTED) {
+    return EFI_UNSUPPORTED;
+  }
+  DEBUG ((DEBUG_ERROR, "[%a] - System proceeded through what should have been a critical failure!\n", __FUNCTION__));
+
+  return EFI_SECURITY_VIOLATION;
+}
+
+/**
+  Unauthorized privileged instruction test.
+  This test will execute platform code to attempt a privileged instruction that should be prevented in SMM
+
+  @retval     EFI_SECURITY_VIOLATION  Since this test is supposed to produce
+                                      a system crash, any sort of return value
+                                      should be considered a security violation.
+
+**/
+EFI_STATUS
+EFIAPI
+SmmMemoryProtectionsTestPrivilegedInstructions (
+  VOID
+  )
+{
+  EFI_STATUS Status;
+  DEBUG ((DEBUG_VERBOSE, "%a()\n", __FUNCTION__));
+
+  // Make sure that the required telemetry/handling is
+  // performed to get accurate test results.
+  EnableExceptionTestMode ();
+
+  DEBUG ((DEBUG_INFO, "[%a] - Attempting unauthorized privileged instruction.\n", __FUNCTION__));
+  Status = TestPrivilegedInstruction();
+  if (Status == EFI_UNSUPPORTED) {
+    return EFI_UNSUPPORTED;
+  }
+  DEBUG ((DEBUG_ERROR, "[%a] - System proceeded through what should have been a critical failure!\n", __FUNCTION__));
+
+  return EFI_SECURITY_VIOLATION;
+}
+
+/**
+  Attempt to write SMM entry point.
+
+  @retval     EFI_SECURITY_VIOLATION  Since this test is supposed to produce
+                                      a system crash, any sort of return value
+                                      should be considered a security violation.
+
+**/
+EFI_STATUS
+EFIAPI
+SmmMemoryProtectionsTestEntryPointAccess (
+  VOID
+  )
+{
+  UINT64 SmmBase;
+  VOID *SmmEntry;
+  DEBUG ((DEBUG_VERBOSE, "%a()\n", __FUNCTION__));
+
+  // Make sure that the required telemetry/handling is
+  // performed to get accurate test results.
+  EnableExceptionTestMode ();
+
+  SmmBase = AsmReadMsr64(MSR_IA32_SMBASE);
```

Comment:
This is Intel centric. Might cause non page fault on other processors.

Reply:
I will move it into the plat library.

Reply:
Done; code moved to the plat library.